src/background_thread.c

```diff
@@ -0,0 +1,453 @@
+		malloc_printf("<jemalloc>: arena %u background thread creation "
+		    "failed (%d).\n", arena_ind, err);
+	}
+	post_reentrancy(tsd);
+
+	if (err == 0) {
+		info->started = true;
+		n_background_threads++;
+	}
+	malloc_mutex_unlock(tsd_tsdn(tsd), &info->mutex);
+
+	return (err != 0);
+}
+
+bool
+background_threads_enable(tsd_t *tsd) {
+	assert(n_background_threads == 0);
+	assert(background_thread_enabled());
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &background_thread_lock);
+
+	VARIABLE_ARRAY(bool, created, ncpus);
+	unsigned i, ncreated;
+	for (i = 0; i < ncpus; i++) {
+		created[i] = false;
+	}
+	ncreated = 0;
+
+	unsigned n = narenas_total_get();
+	for (i = 0; i < n; i++) {
+		if (created[i % ncpus] ||
+		    arena_get(tsd_tsdn(tsd), i, false) == NULL) {
+			continue;
+		}
+		if (background_thread_create(tsd, i)) {
+			return true;
+		}
+		created[i % ncpus] = true;
+		if (++ncreated == ncpus) {
+			break;
+		}
+	}
+
+	return false;
+}
+
+bool
+background_threads_disable(tsd_t *tsd) {
+	assert(!background_thread_enabled());
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &background_thread_lock);
+
+	pre_reentrancy(tsd);
+	for (unsigned i = 0; i < ncpus; i++) {
+		background_thread_info_t *info = &background_thread_info[i];
+		bool wait;
+
+		malloc_mutex_lock(tsd_tsdn(tsd), &info->mutex);
+		if (info->started) {
+			wait = true;
+			info->started = false;
+			pthread_cond_signal(&info->cond);
+		} else {
+			wait = false;
+		}
+		malloc_mutex_unlock(tsd_tsdn(tsd), &info->mutex);
+
+		if (!wait) {
+			continue;
+		}
+		void *ret;
+		if (pthread_join(info->thread, &ret)) {
+			return true;
+		}
+		assert(ret == NULL);
+		n_background_threads--;
+	}
+	post_reentrancy(tsd);
+	assert(n_background_threads == 0);
+
+	return false;
+}
+
+void
+background_thread_prefork(tsdn_t *tsdn) {
+	malloc_mutex_prefork(tsdn, &background_thread_lock);
+	if (background_thread_enabled()) {
+		background_thread_enabled_set(tsdn, false);
+		background_threads_disable(tsdn_tsd(tsdn));
+		/* Enable again to re-create threads after fork. */
+		background_thread_enabled_set(tsdn, true);
+	}
+	assert(n_background_threads == 0);
+	for (unsigned i = 0; i < ncpus; i++) {
+		malloc_mutex_prefork(tsdn, &background_thread_info[i].mutex);
+	}
+}
+
+static void
+background_thread_postfork_init(tsdn_t *tsdn) {
+	if (background_thread_enabled()) {
+		background_threads_enable(tsdn_tsd(tsdn));
+	}
+}
+
+void
+background_thread_postfork_parent(tsdn_t *tsdn) {
+	for (unsigned i = 0; i < ncpus; i++) {
+		malloc_mutex_postfork_parent(tsdn,
+		    &background_thread_info[i].mutex);
+	}
+	background_thread_postfork_init(tsdn);
+	malloc_mutex_postfork_parent(tsdn, &background_thread_lock);
+}
+
+void
+background_thread_postfork_child(tsdn_t *tsdn) {
+	for (unsigned i = 0; i < ncpus; i++) {
+		malloc_mutex_postfork_child(tsdn,
+		    &background_thread_info[i].mutex);
+	}
+	malloc_mutex_postfork_child(tsdn, &background_thread_lock);
+
+	malloc_mutex_lock(tsdn, &background_thread_lock);
```

Comment:
Locking in the postfork handler is non-idiomatic... it seems strange, maybe dangerous?  Can this code be re-ordered to avoid the locking?